File: deck/seed/code/native/kotlin/runtime/variable.kt
// Environment variable runtime. Reached only through the public environment API.
//
// Note the asymmetry the JVM forces: reads come from the process environment (`System.getenv`), which is immutable
// there, so writes go to system properties instead. That is the platform's behaviour, not a choice made here, and it
// is kept in this file so the seed source sees one uniform interface.
object variable {
    fun get(name: String): String = System.getenv(name) ?: ""

    fun set(name: String, value: String) {
        System.setProperty(name, value)
    }

    fun remove(name: String) {
        System.clearProperty(name)
    }

    fun list(): Map<String, String> = System.getenv()

    fun check(name: String): Boolean = System.getenv(name) != null
}

File: deck/seed/code/native/rust/runtime/variable.rs
// Environment variable runtime. Reached only through the public environment API, which is why the platform idioms
// (the `Result` from `env::var`, the borrow at each call) stay here rather than leaking into the seed source.
mod variable {
    use std::collections::HashMap;

    pub fn get(name: String) -> String {
        std::env::var(&name).unwrap_or_default()
    }

    pub fn set(name: String, value: String) {
        std::env::set_var(&name, &value)
    }

    pub fn remove(name: String) {
        std::env::remove_var(&name)
    }

    // the seed hash representation: a reference-counted mutable map
    pub fn list() -> std::rc::Rc<std::cell::RefCell<HashMap<String, String>>> {
        std::rc::Rc::new(std::cell::RefCell::new(std::env::vars().collect()))
    }

    pub fn check(name: String) -> bool {
        std::env::var(&name).is_ok()
    }
}
